test/e2e/storage/testsuites/volumemode.go

```diff
@@ -89,80 +85,16 @@ func (t *volumeModeTestSuite) defineTests(driver TestDriver, pattern testpattern
 
 	init := func() {
 		l = local{}
-		l.ns = f.Namespace
-		l.cs = f.ClientSet
 
 		// Now do the more expensive test initialization.
 		l.config, l.testCleanup = driver.PrepareTest(f)
-
-		fsType := pattern.FsType
-		volBindMode := storagev1.VolumeBindingImmediate
-
-		var (
-			scName             string
-			pvSource           *v1.PersistentVolumeSource
-			volumeNodeAffinity *v1.VolumeNodeAffinity
-		)
-
-		// Create volume for pre-provisioned volume tests
-		l.volume = CreateVolume(driver, l.config, pattern.VolType)
-
-		switch pattern.VolType {
-		case testpatterns.PreprovisionedPV:
-			if pattern.VolMode == v1.PersistentVolumeBlock {
-				scName = fmt.Sprintf("%s-%s-sc-for-block", l.ns.Name, dInfo.Name)
-			} else if pattern.VolMode == v1.PersistentVolumeFilesystem {
-				scName = fmt.Sprintf("%s-%s-sc-for-file", l.ns.Name, dInfo.Name)
-			}
-			if pDriver, ok := driver.(PreprovisionedPVTestDriver); ok {
-				pvSource, volumeNodeAffinity = pDriver.GetPersistentVolumeSource(false, fsType, l.volume)
-				if pvSource == nil {
-					framework.Skipf("Driver %q does not define PersistentVolumeSource - skipping", dInfo.Name)
-				}
-
-				storageClass, pvConfig, pvcConfig := generateConfigsForPreprovisionedPVTest(scName, volBindMode, pattern.VolMode, *pvSource, volumeNodeAffinity)
-				l.sc = storageClass
-				l.pv = framework.MakePersistentVolume(pvConfig)
-				l.pvc = framework.MakePersistentVolumeClaim(pvcConfig, l.ns.Name)
-			}
-		case testpatterns.DynamicPV:
-			if dDriver, ok := driver.(DynamicPVTestDriver); ok {
-				l.sc = dDriver.GetDynamicProvisionStorageClass(l.config, fsType)
-				if l.sc == nil {
-					framework.Skipf("Driver %q does not define Dynamic Provision StorageClass - skipping", dInfo.Name)
-				}
-				l.sc.VolumeBindingMode = &volBindMode
-
-				claimSize := dDriver.GetClaimSize()
-				l.pvc = getClaim(claimSize, l.ns.Name)
-				l.pvc.Spec.StorageClassName = &l.sc.Name
-				l.pvc.Spec.VolumeMode = &pattern.VolMode
-			}
-		default:
-			framework.Failf("Volume mode test doesn't support: %s", pattern.VolType)
-		}
+		l.resource = createVolumeModeTestResource(driver, l.config, pattern)
 	}
 
 	cleanup := func() {
-		if l.pv != nil || l.pvc != nil {
-			By("Deleting pv and pvc")
-			errs := framework.PVPVCCleanup(f.ClientSet, f.Namespace.Name, l.pv, l.pvc)
-			if len(errs) > 0 {
-				framework.Logf("Failed to delete PV and/or PVC: %v", utilerrors.NewAggregate(errs))
-			}
-			l.pv = nil
-			l.pvc = nil
-		}
-
-		if l.sc != nil {
-			By("Deleting sc")
-			deleteStorageClass(f.ClientSet, l.sc.Name)
-			l.sc = nil
-		}
-
-		if l.volume != nil {
-			l.volume.DeleteVolume()
-			l.volume = nil
+		if l.resource != nil {
+			l.resource.cleanupResource()
+			l.resource = nil
```

Comment:
does this need to be set to nil every time?  `l` should be recreated for every test right?

Reply:
Thank you for quick your review.

I think @pohly 's intention was https://github.com/kubernetes/kubernetes/pull/72434#discussion_r257174085 , and I followed the same way.

Reply:
If reviewers find this confusing, perhaps we should remove it everywhere? It's a bit of defensive programming, it shouldn't be absolutely needed.